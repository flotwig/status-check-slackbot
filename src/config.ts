export type Config = {
  port: number
  secret: string
  slackWebhookUrl: string
  targetBranches: string[]
  states: string[]
}

const ensureEnvVar = (name: string): string => {
  const v = process.env[name]

  if (!v) {
    throw new Error(`The ${name} environment variable is required.`)
  }

  return v
}

export const config: Config = {
  port: Number(process.env.PORT) || 3000,
  secret: ensureEnvVar('SECRET'),
  slackWebhookUrl: ensureEnvVar('SLACK_WEBHOOK_URL'),
  targetBranches: (process.env.TARGET_BRANCHES || 'master').split(','),
  states: (process.env.STATES || 'failure,error').split(',')
}
