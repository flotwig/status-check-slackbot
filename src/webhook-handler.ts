import { EventPayloads, WebhookEvent } from '@octokit/webhooks'
import { formatMessage } from './message'
import { Config } from './config'

export const WebhookHandler = ({ states, targetBranches }: Config, send: (message: any) => void) => {
  return ({ payload }: WebhookEvent<EventPayloads.WebhookPayloadStatus>) => {
    if (!states.includes(payload.state)) {
      return
    }

    let branchMatches = false

    for (const targetBranch of targetBranches) {
      if (payload.branches.map(branch => branch.name).includes(targetBranch)) {
        branchMatches = true
        break
      }
    }

    if (!branchMatches) {
      return
    }

    const message = formatMessage(payload)

    send(message)
  }
}
