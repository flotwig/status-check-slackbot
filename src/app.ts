import { Config } from "./config"
import { Webhooks } from '@octokit/webhooks'
import { WebhookHandler } from "./webhook-handler"
import https from 'https'

export const App = (config: Config) => {
  const webhooks = new Webhooks({
    secret: config.secret
  })

  const webhookHandler = WebhookHandler(config, (message) => {
    const req = https.request(config.slackWebhookUrl, {
      method: 'POST'
    })

    req.end(JSON.stringify(message))
  })

  webhooks.on('status', webhookHandler)

  return webhooks.middleware
}
