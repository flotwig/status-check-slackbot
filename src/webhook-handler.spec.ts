import nock from 'nock'
import { App } from './app'
import { Config } from './config'
import { WebhookHandler } from './webhook-handler'
import { WebhookEvent, EventPayloads } from '@octokit/webhooks'

describe('WebhookHandler', function () {
  let config: Config

  beforeEach(function () {
    nock.disableNetConnect()

    config = {
      port: -1,
      secret: 'foo',
      slackWebhookUrl: 'https://slack-webhook-url',
      targetBranches: ['master'],
      states: ['failure', 'error']
    }
  })

  it('posts messages on failures', function (done) {
    const handler = WebhookHandler(config, (message) => {
      expect(message).toMatchSnapshot()
      done()
    })

    const arg = {
      payload: {
        commit: {
          sha: '1234567890abcdef1234567890',
          commit: {
            message: 'commit message',
          },
          author: {
            avatar_url: 'https://author-avatar-url',
            login: 'author-login'
          }
        },
        avatar_url: 'https://avatar-url',
        target_url: 'https://url-to-job',
        context: 'job context',
        description: 'job description',
        state: 'failure',
        branches: [
          { name: 'master' }
        ],
        repository: {
          full_name: 'owner/repo'
        },
      }
    } as unknown as WebhookEvent<EventPayloads.WebhookPayloadStatus>

    handler(arg)
  })
})
