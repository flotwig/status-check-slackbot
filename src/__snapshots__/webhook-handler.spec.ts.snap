// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhookHandler posts messages on failures 1`] = `
Object {
  "blocks": Array [
    Object {
      "text": Object {
        "text": ":x: A status check has failed in \`master\`!

*Commit message*: commit message",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    Object {
      "fields": Array [
        Object {
          "text": "*Job*: <https://url-to-job|job context>",
          "type": "mrkdwn",
        },
        Object {
          "text": "*Message*: <https://url-to-job|job description>",
          "type": "mrkdwn",
        },
        Object {
          "text": "*State*: \`failure\`",
          "type": "mrkdwn",
        },
        Object {
          "text": "*Commit*: <https://github.com/owner/repo/commit/1234567890abcdef1234567890|\`1234567890abcdef\`...> ",
          "type": "mrkdwn",
        },
      ],
      "type": "section",
    },
    Object {
      "elements": Array [
        Object {
          "alt_text": "Status Check Avatar",
          "image_url": "https://avatar-url",
          "type": "image",
        },
        Object {
          "alt_text": "Author Avatar",
          "image_url": "https://author-avatar-url",
          "type": "image",
        },
        Object {
          "text": "*Commit Author*: \`author-login\`",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
}
`;
